homeassistant/components/media_player/gpmdp.py

```diff
@@ -6,42 +6,144 @@
+
+        return
+
+    # pylint: disable=unused-argument
+    def gpmdp_configuration_callback(callback_data):
+        """The actions to do when our configuration callback is called."""
+        while True:
+            msg = json.loads(websocket.recv())
+            if msg['channel'] == 'connect':
+                if msg['payload'] == "CODE_REQUIRED":
+                    websocket.send('{"namespace": "connect",'
+                                   '"method": "connect",'
+                                   '"arguments": ["Home Assistant",'
+                                   ' "' + callback_data.get('pin') + '"]}')
+                    tmpmsg = json.loads(websocket.recv())
+                    if tmpmsg['channel'] == 'time':
+                        _LOGGER.error('Error setting up GPMDP. Please pause'
+                                      'the desktop player and try again.')
+                        break
+                    code = tmpmsg['payload']
+                    if code == 'CODE_REQUIRED':
+                        continue
+                    setup_gpmdp(hass, config, code,
+                                add_devices_callback)
+                    config_from_file(hass.config.path(GPMDP_CONFIG_FILE),
+                                     {"CODE": code})
+                    websocket.send('{"namespace": "connect",'
+                                   '"method": "connect",'
+                                   '"arguments": ["Home Assistant",'
+                                   ' "' + code + '"]}')
+                    websocket.close()
+                    break
+
+    _CONFIGURING['gpmdp'] = configurator.request_config(
+        hass, "GPM Desktop Player", gpmdp_configuration_callback,
+        description=(
+            'Enter the pin that is displayed in the '
+            'Google Play Music Desktop Player.'),
+        submit_caption="Submit",
+        fields=[{'id': 'pin', 'name': 'Pin Code', 'type': 'number'}]
+    )
+
 
+def setup_gpmdp(hass, config, code, add_devices_callback):
+    """Setup gpmdp."""
+    from websocket import create_connection
     name = config.get("name", "GPM Desktop Player")
     address = config.get("address")
+    if not code:
+        websocket = create_connection(("ws://" + address + ":5672"), timeout=1)
+        websocket.send('{"namespace": "connect", "method": "connect",'
+                       '"arguments": ["Home Assistant"]}')
+        request_configuration(hass, config, websocket, add_devices_callback)
+        return
+
+    if 'gpmdp' in _CONFIGURING:
+        configurator = get_component('configurator')
+        configurator.request_done(_CONFIGURING.pop('gpmdp'))
 
     if address is None:
         _LOGGER.error("Missing address in config")
         return False
 
-    add_devices([GPMDP(name, address, create_connection)])
+    add_devices_callback([GPMDP(name, address, code, create_connection)])
+
+
+def config_from_file(filename, config=None):
```

Comment:
I hate this method, please don't copy it 👍 . Use [this](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/notify/html5.py#L73-L100) instead.